addons/website/models/website_visitor.py

```diff
@@ -0,0 +1,177 @@
+
+        for visitor in self:
+            visitor.page_ids = [(6, 0, mapped_data[visitor.id]['page_ids'])] if visitor.id in mapped_data else False
+            visitor.visitor_page_count = mapped_data[visitor.id]['page_count'] if visitor.id in mapped_data else False
+            visitor.page_ids_count = len(mapped_data[visitor.id]['page_ids']) if visitor.id in mapped_data else 0
+
+    @api.depends('last_connection_datetime')
+    def _compute_time_statistics(self):
+        query = """ SELECT id, last_connection_datetime
+                    FROM website_visitor
+                    WHERE id in %s"""
+        self.env.cr.execute(query, [tuple(self.ids)])
+        results = self.env.cr.dictfetchall()
+        mapped_data = {result['id']: result['last_connection_datetime'] for result in results}
+
+        for visitor in self:
+            last_connection_date = mapped_data[visitor.id]
+            visitor.time_since_last_action = self._format_last_action_time((datetime.now() - last_connection_date).total_seconds())
+            visitor.is_connected = (datetime.now() - last_connection_date) < timedelta(minutes=5)
+
+    def _format_last_action_time(self, seconds):
+        if seconds / 60 <= 1:
+            return "%s sec ago" % round(seconds)
+        elif seconds / (60 * 60) <= 1:
+            return "%s min ago" % round(seconds / 60)
+        elif seconds / (60 * 60 * 24) <= 1:
+            return "%s h ago" % round(seconds / (60*60))
+        else:
+            nbr_days = round(seconds / (60 * 60 * 24))
+            return "%s day%s ago" % (nbr_days, '' if nbr_days == 1 else 's')
+
+    def _encode(self, visitor_id):
+        return "%d-%s" % (visitor_id, self._get_visitor_hash(visitor_id))
+
+    def _get_visitor_hash(self, visitor_id):
+        return md5(("%s%s" % (visitor_id, self.env['ir.config_parameter'].sudo().get_param('database.secret'))).encode('utf-8')).hexdigest()
+
+    def _get_visitor_from_request(self):
+        if not request:
+            return None
+        visitor = self.env['website.visitor']
+        cookie_content = request.httprequest.cookies.get('visitor_id')
+        if cookie_content and '-' in cookie_content:
+            visitor_id, visitor_hash = cookie_content.split('-', 1)
+            if _consteq(visitor_hash, self._get_visitor_hash(visitor_id)):
+                return visitor.sudo().with_context(active_test=False).search([('id', '=', visitor_id)])  # search to avoid having to call exists()
+        return visitor
+
+    def _handle_webpage_dispatch(self, response, is_webpage=False):
+        """ Called on dispatch. This will create a website.visitor if the http request object
+        is a tracked website page. Only on tracked page to avoid having too much operations done on every page
+        or other http requests.
+        Note: The side effect is that the last_connection_datetime is updated ONLY on tracked pages."""
+        website_page = is_webpage and response.qcontext['main_object']
```

Comment:
Is this necessary?
I feel like your test above already handles that condition (`_is_webpage_response`).